### DISCUSSION TITLE
Occasional Debug Shutdown

### DISCUSSION THREAD

**Discussion by rrirower** [rokucommunity/brighterscript#1035] [Q&A]
Any thoughts on how to avoid the occasional debug shutdown while in the middle of examining code?

**Comment by TwitchBronBron** [comment]
There are several different factors at play here. 
1. The most common shutdown is due to the roku device just giving up and quitting. We haven't figured out why, but there are just random times where it'll decide to shut everything down. I know that they have some fairly aggressive thread detection software running on devices, so perhaps our debug sessions make the device think it's running malicious software so it kills it? There's not much we can do about this scenario
2. When using the debug protocol, we have a few timing bugs where the session will start, but then after 20-30 seconds it'll disconnect. That's due to the device not accepting our connection request when we send it. We have polling on to try and keep reconnecting, but sometimes it just gets in an unrecoverable state. We do have a workaround for this in progress, which should hopefully entirely eliminate this issue.
3. The telnet debugger is quite touchy. We are basically running a giant text parsing state machine. There are many situations where we get text output from the device that we hadn't expected, and it totally breaks our state machine. This won't necessarily cause the debug shutdown from the device side, but getting into an incorrect state in our state machine might cause us to throw exceptions when we think we're in state a, but are actually in state b so we do some type of runtime operation that crashes the nodejs process.

That's all I can currently think of. I'm happy to help figure out solutions to anything that we have control over, but we'll need some fairly consistent reproducible steps (or at least good logs showing what went wrong) in order to know where to begin to fix it.
